File: JAVA/Fence.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Stack;
import java.util.Vector;


public class Fence {
	
	private static Vector<Integer> height;

	public static void main (String[] args) {
		try {
			BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
			BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
			int cases = Integer.parseInt(reader.readLine());
			
			while (cases-- > 0) {
				int numOfFence = Integer.parseInt(reader.readLine());
				height = new Vector<Integer>(numOfFence);
				String[] input = reader.readLine().split(" ");
				for (int i = 0; i < numOfFence; i++) height.add(Integer.parseInt(input[i]));
				writer.append(String.valueOf(maxRectSize())); writer.append("\n");
			}
			writer.flush();
			writer.close();
			reader.close();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	public static int maxRectSize() {
		Stack<Integer> remaining = new Stack<Integer>();
		height.add(0);
		int ret = 0;
		for (int i = 0; i < height.size(); i++) {
			while (!remaining.empty() && height.get(remaining.peek()) >= height.get(i)) {
				int j = remaining.pop();
				int width = -1;
				if (remaining.empty()) width = i;
				else width = i - remaining.peek() - 1;
				ret = Math.max(ret, height.get(j) * width);
			}
			remaining.push(i);
		}
		return ret;
	}
}

File: CPP/Fence.cpp
#include <cstdio>
#include <stack>
#include <vector>
using namespace std;

vector<int> height;

int max(int a, int b) { return a > b ? a : b;}

int maxRectSize() {
	int ret = 0;
	stack<int> remaining;
	height.push_back(0);
	for (int i = 0; i < height.size(); i++) {
		while (!remaining.empty() && height[remaining.top()] > height[i]) {
			int j = remaining.top();
			int width = -1;
			remaining.pop();
			if (remaining.empty()) width = 1;
			else width = i - remaining.top() - 1;
			ret = max(ret, width * height[j]);
		}
		remaining.push(i);
	}
	return ret;
}

int main() {
	int cases; scanf("%d", &cases);

	while (cases--) {
		int numOfFence; scanf("%d", &numOfFence);
		for (int i = 0; i < numOfFence; i++) {
			int value; scanf("%d", &value);
			height.push_back(value);
		}
		printf("%d\n", maxRectSize());
		height.clear();
	}
}
